teos/teosd.py

```diff
@@ -68,171 +56,291 @@ def get_config(command_line_conf, data_dir):
+        # https://github.com/talaia-labs/python-teos/pull/198
+        self.internal_api.rpc_server.start()
+        logger.info(f"Internal API initialized. Serving at {INTERNAL_API_ENDPOINT}")
+
+        # Start the public API server
+        api_endpoint = f"{self.config.get('API_BIND')}:{self.config.get('API_PORT')}"
+        if self.config.get("WSGI") == "gunicorn":
+            # FIXME: We may like to add workers depending on a config value
+            self.api_proc = subprocess.Popen(
+                [
+                    "gunicorn",
+                    f"--bind={api_endpoint}",
+                    f"teos.api:serve(internal_api_endpoint='{INTERNAL_API_ENDPOINT}', "
+                    f"endpoint='{api_endpoint}', min_to_self_delay='{self.config.get('MIN_TO_SELF_DELAY')}', "
+                    f"log_file='{self.config.get('LOG_FILE')}')",
+                ]
             )
-            db_manager = AppointmentsDBM(config.get("APPOINTMENTS_DB_PATH"))
-            responder = Responder(db_manager, gatekeeper, carrier, block_processor)
-            watcher = Watcher(
-                db_manager,
-                gatekeeper,
-                block_processor,
-                responder,
-                sk,
-                config.get("MAX_APPOINTMENTS"),
-                config.get("LOCATOR_CACHE_SIZE"),
+        else:
+            self.api_proc = multiprocessing.Process(
+                target=api.serve,
+                kwargs={
+                    "internal_api_endpoint": INTERNAL_API_ENDPOINT,
+                    "endpoint": api_endpoint,
+                    "min_to_self_delay": self.config.get("MIN_TO_SELF_DELAY"),
+                    "log_file": self.config.get("LOG_FILE"),
+                    "auto_run": True,
+                },
             )
+            self.api_proc.start()
 
-            # Create the chain monitor and start monitoring the chain
-            chain_monitor = ChainMonitor(
-                watcher.block_queue, watcher.responder.block_queue, block_processor, bitcoind_feed_params
-            )
+    def handle_signals(self, signum, frame):
+        """Handles signals by initiating a graceful shutdown."""
+        logger.debug(f"Signal {signum} received. Stopping")
 
-            watcher_appointments_data = db_manager.load_watcher_appointments()
-            responder_trackers_data = db_manager.load_responder_trackers()
+        self.stop_command_event.set()
 
-            if len(watcher_appointments_data) == 0 and len(responder_trackers_data) == 0:
-                logger.info("Fresh bootstrap")
+    def teardown(self):
+        """Shuts down all services and closes the DB, then exits. This method does not return."""
+        logger.info("Terminating public API")
 
-                watcher.awake()
-                watcher.responder.awake()
+        # Stop the public API first
+        if self.config.get("WSGI") == "gunicorn":
```

Comment:
I guess this can also be checked with `isinstance`, so the comments won't be necessary.

Reply:
Changed to check using `isinstance` in f3135e8; I also added an `else` that raises an error, to be explicit about the types.